packages/server/utils/atlassian/jiraIssues.ts

```diff
@@ -0,0 +1,121 @@
+import getRedis from '../getRedis'
+import AtlassianManager, {RateLimitError, JiraIssueRaw} from 'parabol-client/utils/AtlassianManager'
+import ms from 'ms'
+import sleep from 'parabol-client/utils/sleep'
+import stringify from 'fast-json-stable-stringify'
+
+const ISSUE_TTL_MS = ms('2d')
+const MAX_RETRIES = 4
+
+type StoreAndNetworkRequests = RequestsInit | CacheHit | CacheMiss | CacheError
+
+interface RequestsInit {
+  firstToResolve: undefined | ((value: unknown) => void)
+  cachedIssue: undefined
+}
+interface CacheHit {
+  firstToResolve: (value: unknown) => void
+  cachedIssue: string
+}
+interface CacheMiss {
+  firstToResolve: (value: unknown) => void
+  cachedIssue: null
+}
+type CacheError = RequestsInit
+
+/* helper fn to only do jira request side effects and return nothing */
+const _getIssueFromJira = async (
+  manager: AtlassianManager,
+  cloudId: string,
+  issueKey: string,
+  pushUpdateToClient: (
+    issue: JiraIssueRaw
+  ) => void /* cb invoked when we get a fresher jira issue */,
+  extraFieldIds: string[],
+  redisKey: string,
+  requests: StoreAndNetworkRequests, // resolve originally created promise in outer fn
+  retryNo = 0
+) => {
+  // get issue from external network (jira api)
+  manager.getIssue(cloudId, issueKey, extraFieldIds).then(async (res) => {
+    if (res instanceof RateLimitError) {
+      const {retryAt} = res
+      const delay = Math.max(0, retryAt.getTime() - Date.now())
+      if (delay > ms('10s') || retryNo > MAX_RETRIES) return
```

Comment:
if the delay is > 10 seconds, when does `requests.firstToResolve` ever get called? i think this would cause a mem leak because the promise would never resolve?

Reply:
you are totally correct. i overlooked this, thank you for catching it